sdk/src/logging/FileLogDestination.cpp

```diff
@@ -115,21 +278,227 @@ struct FileLogDestination::Impl
+      if (LogOptions.getMaxRotations() < 1000 && logFileNumber > 1000)
+         return false;
+
+      FilePath rotatedLogFile = getLogRotationFile(logFile, logFileNumber);
+
+      // Recursively rotate each file in the numeric sequence
+      // (.1.log becomes .2.log, .2.log becomes .3.log, etc)
+      if (rotatedLogFile.exists())
+      {
+         if (!rotateLogFileImpl(rotatedLogFile, logFileNumber))
+            return false;
+      }
+
+      if (LogOptions.getMaxRotations() > 0)
+      {
+         // If we are over the maximum configured allowed rotated log files, delete this
+         // file instead of rotating it
+         if (logFileNumber > LogOptions.getMaxRotations())
+         {
+            logFile.remove();
+            return true;
+         }
+      }
+
+      if (LogOptions.getDeletionDays() > 0)
+      {
+         // Check to see if the logfile is stale
+         // If so, we will delete it instead of rotating it
+         std::time_t lastWrite = logFile.getLastWriteTime();
+         DateTime lastWriteTime = lastWrite != 0 ? DateTime(lastWrite) :
+                                                                   DateTime();
+         DateTime now;
+          if ((now - lastWriteTime).getHours() > TimeDuration(24).getHours() * LogOptions.getDeletionDays())
+         {
+            logFile.remove();
+            return true;
+         }
+      }
+
+      // Now do the current rotation for this file
+      Error error = logFile.move(rotatedLogFile);
       if (error)
          return false;
 
       return true;
    }
 
+   void chownLogs(const FilePath& logFile,
+                  const launcher_plugins::system::User& user,
+                  int logFileNumber = 0)
+   {
+      logFileNumber += 1;
+
+      // Defensive break in case we somehow find ourselves in a situation where
+      // we have an unreasonable amount of log files to chown
+      if (LogOptions.getMaxRotations() < 1000 && logFileNumber > 1000)
+         return;
+
+      FilePath rotatedLogFile = getLogRotationFile(logFile, logFileNumber);
+
+      // Recursively chown each file in the numeric sequence
+      if (rotatedLogFile.exists())
+         chownLogs(rotatedLogFile, user, logFileNumber);
+
+      // Now do the current chown for this file
+      logFile.changeOwnership(user);
+   }
+
+   void setLogDirOwner(const launcher_plugins::system::User& user)
+   {
+      LogOptions.getDirectory().changeOwnership(user);
+   }
+
+   DateTime getTimestampFromLogLine(const std::string& line)
+   {
+      // Empty line means the file hasn't actually been written to yet, so bail early
+      if (line.empty())
+         return DateTime();
+
+      DateTime time;
+      std::string timeStr;
+
+      // Check for time in human readable log format first
+      size_t pos = line.find(' ');
+      if (pos != std::string::npos)
+      {
+         timeStr = line.substr(0, pos);
+         if (launcher_plugins::system::DateTime::fromString(timeStr,"",time))
```

Comment:
```suggestion
         if (DateTime::fromString(timeStr,"",time))
```

Don't need the extra namespaces specified here.